### PR TITLE
Add Ruby 3.4.0-preview2 support for CircleCI tests

### PR DESCRIPTION
**Pull request by sarahchen6**
**What does this PR do?**
This PR adds "3.4" versions to the `config.yml` file under `.circleci` everywhere there is a comment to `# ADD NEW RUBIES HERE`.

**Motivation:**
Though the official Ruby 3.4 is not yet released, we can temporarily add Ruby 3.4.0-preview2. This allows us to start testing with Ruby 3.4 in our CI and spot -> address issues early. After Ruby 3.4 is officially released post-Dec 25, 2024, we can replace the preview with it.

**Change log entry**
No change log entry for now. Thinking to have a "Ruby 3.4.0-preview2 support added" update after it is fully supported.

**Additional Notes:**
This PR is one step toward full support and only addresses CircleCI workflow tests. It acts alongside #4038 and #4039 .

**How to test the change?**
This change can be tested in CI.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -90,3 +86,15 @@ end
...
+# `1.17.0` provides broken RBS type definitions
+# https://github.com/ffi/ffi/blob/master/CHANGELOG.md#1170rc1--2024-04-08
+#
+# TODO: Remove this once the issue is resolved: https://github.com/ffi/ffi/issues/1107
+gem 'ffi', '~> 1.16.3', require: false
+
+if RUBY_VERSION.start_with?('3.4.')
+  gem 'strscan',
+    git: 'https://github.com/ruby/strscan',
+    ref: '041b15df4ccc067deabd85fd489b2c15961d0e2f'
```
Comment (p-datadog) [inline Gemfile:99]: Can this get a comment explaining why we need this particular version?

File: appraisal/ruby-3.4.rb [hunk 2]
```diff
@@ -94,7 +94,7 @@
   gem 'makara', '>= 0.6.0.pre' # Ruby 3 requires >= 0.6.0, which is currently in pre-release: https://rubygems.org/gems/makara/versions
   gem 'mysql2', '>= 0.5.3', platform: :ruby
   gem 'pg', platform: :ruby
-  gem 'sqlite3', '>= 1.4.2', platform: :ruby
+  gem 'sqlite3', '~> 1.3', '>= 1.3.6', platform: :ruby
```
Comment (p-datadog) [inline appraisal/ruby-3.4.rb:97]: Is this still necessary?

File: Rakefile [hunk 3]
```diff
@@ -55,6 +55,11 @@ namespace :test do
         end
       end
 
+      sh('gem -v')
+      sh('bundler -v')
+      sh('gem list')
+      sh('bundle show')
```
Comment (p-datadog) [inline Rakefile:61]: This should be removed before merging (as soon as CI is green).

File: appraisal/ruby-3.4.rb (hunk 2 above)
Comment (sarahchen6) [reply appraisal/ruby-3.4.rb:97]: Right. Let me adjust this.

Comment (sarahchen6) [reply appraisal/ruby-3.4.rb]: It is still necessary..

**Comment by p-datadog** [comment]
`Rack::Handler` was also deleted in 3.1.0. The removal is mentioned in https://github.com/rack/rack/blob/main/UPGRADE-GUIDE.md but it looks like some/all of those changes were made in 3.1.0 and not in 3.0 as advertised.

**Comment by p-datadog** [comment]
The following makes the test pass on ruby-3.4 config:

```
diff --git a/spec/datadog/tracing/contrib/suite/integration_spec.rb b/spec/datadog/tracing/contrib/suite/integration_spec.rb
index eaf37eb1f..eaa9ab0ba 100644
--- a/spec/datadog/tracing/contrib/suite/integration_spec.rb
+++ b/spec/datadog/tracing/contrib/suite/integration_spec.rb
@@ -8,7 +8,7 @@ require 'semantic_logger'
 
 require 'rack'
 # `Rack::Handler::WEBrick` was extracted to the `rackup` gem in Rack 3.0
-require 'rackup' if Gem::Version.new(Rack::RELEASE) >= Gem::Version.new('3')
+require 'rackup/handler/webrick' if Gem::Version.new(Rack::RELEASE) >= Gem::Version.new('3')
 require 'webrick'
 
 RSpec.describe 'contrib integration testing', :integration do
@@ -164,7 +164,7 @@ RSpec.describe 'contrib integration testing', :integration do
             end
           end.to_app
 
-          server.mount '/', Rack::Handler::WEBrick, app
+          server.mount '/', Rackup::Handler::WEBrick, app
 
           @thread = Thread.new { server.start }
           try_wait_until { started }
```

I haven't checked if this would also work on rack 3.0, and I am quite certain this will NOT work on rack 2 and lower therefore the change needs to be conditional on rack version.

**Comment by ivoanjo** [comment]
Slight suggestion: This PR is getting so big at this point that GitHub literally does not show me the complete list of files changed when I go on the diff tab.

May be worth splitting some of the changes into smaller PRs that could go in already? E.g. maybe a separate PR for the ffi/webrick stuff, that doesn't even add 3.4 yet, etc?

**Comment by sarahchen6** [comment]
This PR was split up. See #4114 .

**Comment by ivoanjo** [comment]
> This PR was split up. See https://github.com/DataDog/dd-trace-rb/pull/4114 .

Thank you :pray: :D

**Pull request closed without merging**